apidef/oas/operation.go

```diff
@@ -0,0 +1,157 @@
+package oas
+
+import (
+	"strings"
+
+	"github.com/TykTechnologies/tyk/apidef"
+	"github.com/getkin/kin-openapi/openapi3"
+)
+
+type Operations map[string]*Operation
+
+type Operation struct {
+	Allow                *Allowance `bson:"allow,omitempty" json:"allow,omitempty"`
+	Block                *Allowance `bson:"block,omitempty" json:"block,omitempty"`
+	IgnoreAuthentication *Allowance `bson:"ignoreAuthentication,omitempty" json:"ignoreAuthentication,omitempty"`
+}
+
+const (
+	allow                AllowanceType = 0
+	block                AllowanceType = 1
+	ignoreAuthentication AllowanceType = 2
+)
+
+type AllowanceType int
+
+func (s *OAS) fillPathsAndOperations(ep apidef.ExtendedPathsSet) {
+	if s.Paths == nil {
+		s.Paths = make(openapi3.Paths)
+	}
+
+	s.fillAllowance(ep.WhiteList, allow)
+	s.fillAllowance(ep.BlackList, block)
+	s.fillAllowance(ep.Ignored, ignoreAuthentication)
+
+	if len(s.Paths) == 0 {
+		s.Paths = nil
+	}
+}
+
+func (s *OAS) extractPathsAndOperations(ep *apidef.ExtendedPathsSet) {
+	tykOperations := s.getTykOperations()
+	if len(tykOperations) == 0 {
+		return
+	}
+
+	for id, tykOp := range tykOperations {
+		for path, pathItem := range s.Paths {
+			for method, operation := range pathItem.Operations() {
+				if id == operation.OperationID {
+					tykOp.extractAllowanceTo(ep, path, method, allow)
+					tykOp.extractAllowanceTo(ep, path, method, block)
+					tykOp.extractAllowanceTo(ep, path, method, ignoreAuthentication)
+				}
+			}
+		}
+	}
+}
+
+func (s *OAS) fillAllowance(endpointMetas []apidef.EndPointMeta, typ AllowanceType) {
+	for _, em := range endpointMetas {
+		operationID := s.getOperationID(em.Path, em.Method)
+		operation := s.GetTykExtension().getOperation(operationID)
+
+		var allowance *Allowance
+
+		switch typ {
+		case block:
+			if operation.Block == nil {
+				operation.Block = &Allowance{}
+			}
+
+			allowance = operation.Block
+		case ignoreAuthentication:
+			if operation.IgnoreAuthentication == nil {
+				operation.IgnoreAuthentication = &Allowance{}
+			}
+
+			allowance = operation.IgnoreAuthentication
+		default:
+			if operation.Allow == nil {
+				operation.Allow = &Allowance{}
+			}
+
+			allowance = operation.Allow
+		}
+
+		allowance.Fill(em)
```

Comment:
A field value of `em` (apidef.EndPointMeta) is being modified, but not passed as a pointer value. Should it be 'fillAllowance' responsibility to populate .Block, .IgnoreAuthentication, .Allow? Or should we have something more like:

```go
switch typ {
case block:
    allowance = operation.Block.Fill(em)
case ignoreAuthentication:
    allowance = operation.IgnoreAuthentication.Fill(em)
...
```
Fill() would return *Allowance, and could work on a nil value for the receiver to create a new Allowance object as needed